### ISSUE TITLE
Batched GEMM is not lane-symmetric for small shapes on ROCm (MI350X): identical batch rows give different outputs

### ISSUE THREAD

**Issue by zjin-lcf** [pytorch/pytorch#192420]
### 🐛 Describe the bug

On AMD Instinct MI350X (gfx950, ROCm 7.2), a **batched GEMM produces bit-different results across batch lanes for bit-identical input rows** at small sequence lengths. Two rows that are `torch.equal`-identical on input give non-identical outputs from `nn.Linear` / `torch.matmul`.

This breaks batch-invariance / lane-symmetry that user code (and greedy decoding) reasonably relies on: identical prompts placed in different batch positions should produce identical logits. The ~1e-6 lane difference is tiny in isolation, but in a deep autoregressive decoder it amplifies layer-over-layer and eventually flips a near-tie `argmax`, causing two identical prompts in the same batch to generate different sequences.

Key observations:
- The default (`hipblaslt`) and `hipblas` backends are **not** lane-symmetric for the small shape; the **`ck` (Composable Kernel)** backend **is** (bitwise-equal lanes).
- Computing each lane as its own `batch=1` GEMM is bitwise-equal — so it is the batched/tiled reduction across the leading dim, not the math.
- `torch.use_deterministic_algorithms(True)` does **not** fix it.
- It reproduces even in **fp32-accumulated** matmul, so it is not a bf16 rounding artifact — it is kernel tiling.
- Large `S` (e.g. 512) is lane-symmetric, i.e. a different, symmetric kernel is selected for larger shapes.

### Minimal repro

```python
import torch

dev = "cuda"
print("device:", torch.cuda.get_device_name(0),
      "| hip", getattr(torch.version, "hip", None),
      "| torch", torch.__version__)
print("default BLAS backend:", torch.backends.cuda.preferred_blas_library())

S, IN, OUT = 24, 1024, 2048
torch.manual_seed(0)
row = torch.randn(1, S, IN, device=dev, dtype=torch.bfloat16)
x = torch.cat([row, row], dim=0)          # two bit-identical rows
assert torch.equal(x[0], x[1])
lin = torch.nn.Linear(IN, OUT, bias=False).to(dev, torch.bfloat16)

def report(tag, y):
    eq = bool(torch.equal(y[0], y[1]))
    d = (y[0].float() - y[1].float()).abs().max().item()
    print(f"  {tag:38s} row0==row1 {str(eq):5s}  maxabs={d:.3e}")

print("\nbatched nn.Linear, (2,%d,%d)->(2,%d,%d):" % (S, IN, S, OUT))
for be in ["default", "hipblas", "hipblaslt", "ck"]:
    torch.backends.cuda.preferred_blas_library(be)
    with torch.no_grad():
        y = lin(x)
    report(f"backend={be}", y)
torch.backends.cuda.preferred_blas_library("default")

with torch.no_grad():
    y0 = lin(x[0:1]); y1 = lin(x[1:2])
print("\nper-lane batch=1: lane0==lane1:", bool(torch.equal(y0[0], y1[0])))

torch.use_deterministic_algorithms(True)
with torch.no_grad():
    report("use_deterministic_algorithms(True)", lin(x))
torch.use_deterministic_algorithms(False)

with torch.no_grad():
    report("fp32-accumulated matmul", torch.matmul(x.float(), lin.weight.t().float()))

torch.manual_seed(0)
big = torch.randn(1, 512, IN, device=dev, dtype=torch.bfloat16)
with torch.no_grad():
    report("nn.Linear S=512", lin(torch.cat([big, big], dim=0)))
```

### Output on MI350X

```
device: AMD Instinct MI350X | hip 7.2.53211 | torch 2.10.0+git94c6e04
default BLAS backend: _BlasBackend.Cublaslt

batched nn.Linear, (2,24,1024)->(2,24,2048):
  backend=default                        row0==row1 False  maxabs=3.815e-06
  backend=hipblas                        row0==row1 False  maxabs=3.815e-06
  backend=hipblaslt                      row0==row1 False  maxabs=3.815e-06
  backend=ck                             row0==row1 True   maxabs=0.000e+00

per-lane batch=1: lane0==lane1: True
use_deterministic_algorithms(True)     row0==row1 False  maxabs=3.815e-06
fp32-accumulated matmul                row0==row1 False  maxabs=7.153e-07

nn.Linear S=512                        row0==row1 True   maxabs=0.000e+00
```

### Expected behavior

Batched GEMM should be lane-symmetric: for `x` with `torch.equal(x[0], x[1])`, the output should satisfy `torch.equal(y[0], y[1])`. This holds for the `ck` backend, for per-lane `batch=1` GEMMs, and for large `S`, but not for the default `hipblaslt`/`hipblas` kernels at small shapes. (`ck` is not a general workaround: it lacks kernels for many shapes and raises `device_gemm ... does not support this GEMM problem` in an MoE forward.)

### Impact

In a real model (Qwen3-Omni talker, a 20-layer autoregressive MoE decoder), two bit-identical prompts batched together diverge starting at `layers.0.self_attn.q_proj` (a plain `nn.Linear`, maxabs ~3e-5), amplify to ~2.2 by layer 20, and flip the greedy `argmax`, producing different generated sequences for identical inputs. Forcing a lane-symmetric `F.linear` removes the attention-projection divergence; the next divergence then appears in the MoE expert GEMM (also `nn.functional.linear`, on a 2D gathered-token tensor), confirming the same asymmetry across every dense matmul path.

### Versions

- GPU: AMD Instinct MI350X (gfx950)
- ROCm/HIP: 7.2.53211
- PyTorch: 2.10.0+git94c6e04
- dtype: bfloat16 (also reproduces in fp32)

cc @ezyang @gchanan @kadeng @msaroufim @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @mruberry @kurtamohler @jianyuh @nikitaved @walterddr @Lezcano

**Comment by naromero77amd** [comment]
@zjin-lcf I was able to reproduce this behavior on ROCm 7.14 on an MI200 and with different data types. I think we need to get hipBLASLt team involved here.
